Heat gui

drchrislewis opened the pull request:
This provides an interface to the heat method, and therefore depends on this repo:
https://mrtd-gitlab.dyn.datasys.swri.edu/r6029/heat.git branch heat1_lha, also in review
It allows one to generate paths using the heat method by first drawing the sources. The interface is functional, but incomplete. For example, one may not segment, nor select a region prior to defining the source. This is a PPPI. 
Also, the draw a path directly does not work, but one may accomplish the same thing by setting the line spacing very high. This results in a shortest on mesh path between each pair of polyline points provided.

Commit:
just checking in to provide backup

Commit:
attempting to delete ~ files accidentaly added

Commit:
saving before trip so I can work on another machine

Commit:
Georgia heat

Commit:
integrated heat method, but heat method's paths are not smoothed, nor sub-sampled to desired point spacing

Commit:
working interface to heat method, but not if you segment first.

Commit:
just saving so I can switch branches to test that I didn't break anything

Commit:
changed color of path polyline to blue to distinguish it from the area polyline. removed an unused button

marip8 commented on opp_msgs/CMakeLists.txt:
```diff
@@ -29,7 +29,9 @@ add_message_files(
   FILES
-    GetROISelection.srv
+      GetROISelection.srv
+      GetPathSelectionMesh.srv
+      GetPathSelectionCloud.srv
```
You might consider combining these messages if you don't need access to the triangle information from the mesh

drchrislewis replied:
Connor is tasked with adding the cloud rather than mesh implementation. I just added the srv definition in anticipation of that implementation. We definitely use the triangle info from the mesh. Once the cloud version is working, it may be possible to treat meshes as clouds, but not yet.

marip8 commented on opp_msgs/srv/GetPathSelection.srv:
```diff
@@ -0,0 +1,12 @@
+# Request
+
+# The set of points from which the ROI
+# will be selected
+sensor_msgs/PointCloud2 input_cloud
+
+---
+
+# Response
+bool success
+string message
+int32[] cloud_indices
```
This service message isn't built by the CMakeLists; it should be removed or the the CMakeLists should be amended

marip8 commented on opp_path_selection/include/opp_path_selection/filter.h:
```diff
@@ -0,0 +1,44 @@
+
+}  // end namespace opp_path_selection
+
+#include "opp_path_selection/filter_impl.h"
```
I would suggest not including the implementation in the header. Rather, build a library with specific implementations of the template (using the `PCL_INSTANTIATE` macro). If you need a type that isn't already built, I would prefer to include the implementation directly.

drchrislewis replied:
This is a direct copy of the area_selector code with the changes necessary to create a path rather than an area. I would have never ever created the xxx_impl.h otherwise. I do not understand what you are suggesting. I'll take this one off line.

marip8 commented on opp_path_selection/include/opp_path_selection/filter_impl.h:
```diff
@@ -0,0 +1,76 @@
+#ifndef OPP_PATH_SELECTION_FILTER_IMPL_H
+#define OPP_PATH_SELECTION_FILTER_IMPL_H
+
+#include "opp_path_selection/filter.h"
```
This technically introduces a circular dependency since your header includes this file. This is handled by the header guards, but isn't necessarily great. The above comment would address this

drchrislewis replied:
We should fix both packages as you suggest. opp_area_selection and opp_path_selection.

marip8 commented on opp_path_selection/include/opp_path_selection/path_selector.h:
```diff
@@ -0,0 +1,105 @@
+#include <boost/optional.hpp>
+#include <eigen3/Eigen/Core>
+
+#include <geometry_msgs/Point.h>
+#include <geometry_msgs/PoseStamped.h>
+#include <pcl_ros/point_cloud.h>
+#include <sensor_msgs/PointCloud2.h>
+#include <shape_msgs/Mesh.h>
+#include "opp_path_selection/path_selector_parameters.h"
+
+#include <boost/property_map/property_map.hpp>
+#include <boost/graph/graph_traits.hpp>
+#include <boost/graph/adjacency_list.hpp>
+#include <boost/graph/dijkstra_shortest_paths.hpp>
+#include <boost/graph/breadth_first_search.hpp>
+#include <boost/graph/graph_utility.hpp>
+#include <boost/pending/indirect_cmp.hpp>
+#include <boost/range/irange.hpp>
```
Move as many headers as possible to the source file

drchrislewis replied:
Good idea.

marip8 commented on opp_path_selection/include/opp_path_selection/path_selector_parameters.h:
```diff
@@ -0,0 +1,31 @@
+#ifndef OPP_PATH_SELECTION_PATH_SELECTOR_PARAMETERS_H
+#define OPP_PATH_SELECTION_PATH_SELECTOR_PARAMETERS_H
+
+#include <limits>
```
Unnecessary include?

marip8 replied:
Also, you might consider combining this with the path selector file; it doesn't seem necessary to put this in a separate file

marip8 commented on opp_path_selection/src/path_selection_artist.cpp:
```diff
@@ -0,0 +1,487 @@
+const static std::string CLICKED_POINT_TOPIC = "clicked_point";
+const static std::string CLEAR_PATH_POINTS_SERVICE = "clear_path_selection_points";
+const static std::string COLLECT_PATH_POINTS_SERVICE = "collect_path_selection_points";
+const static std::string path_selection_config_file = ros::package::getPath("opp_path_selection") + "/config/"
```
This seems like something that should be passed in to a function or constructor as a string argument rather than burying a hard-coded file path in a source file

marip8 commented on opp_path_selection/src/path_selection_artist.cpp:
```diff
@@ -0,0 +1,487 @@
+  return visuals;
+}
+
+bool pclToShapeMsg(const pcl::PolygonMesh& pcl_mesh, shape_msgs::Mesh& mesh_msg)
```
I'm not sure if it's worthwhile to do now, but it would probably be much easier to use the `pcl_msgs/PolygonMesh` representation of a mesh instead of `shape_msgs/Mesh`

marip8 commented on opp_path_selection/src/path_selection_artist.cpp:
```diff
@@ -0,0 +1,487 @@
+                                 const std::string& sensor_frame)
+  : nh_(nh), world_frame_(world_frame), sensor_frame_(sensor_frame)
+{
+  marker_pub_ = nh_.advertise<visualization_msgs::MarkerArray>(MARKER_ARRAY_TOPIC, 1, false);
```
Typically latching publishers (i.e. third argument is `true`) is more convenient for seeing displays in Rviz because you don't lose the visualization after toggling it. When a publisher is not latched a new message must be published before it can be re-visualized

marip8 commented on opp_path_selection/src/path_selection_artist.cpp:
```diff
@@ -0,0 +1,487 @@
+  : nh_(nh), world_frame_(world_frame), sensor_frame_(sensor_frame)
+{
+  marker_pub_ = nh_.advertise<visualization_msgs::MarkerArray>(MARKER_ARRAY_TOPIC, 1, false);
+  listener_.reset(new tf::TransformListener(nh_));
```
Prefer to use `tf2`; also there's probably no reason for this to be a pointer

marip8 commented on opp_path_selection/src/path_selection_artist.cpp:
```diff
@@ -0,0 +1,487 @@
+  marker_pub_ = nh_.advertise<visualization_msgs::MarkerArray>(MARKER_ARRAY_TOPIC, 1, false);
+  listener_.reset(new tf::TransformListener(nh_));
+
+  if (!listener_->waitForTransform(sensor_frame_, world_frame_, ros::Time(0), ros::Duration(TF_LOOKUP_TIMEOUT)))
+  {
+    ROS_ERROR("Transform lookup from %s to %s timed out", sensor_frame_.c_str(), world_frame_.c_str());
+    throw std::runtime_error("Transform lookup timed out");
+  }
```
Why not just handle the failure case when you go to look up the transformation? Maybe the transformation will exist at that time. I've personally stopped doing checks like this because it's easier to handle the failure where it happens rather than trying to preempt it with a check like this

marip8 commented on opp_path_selection/src/path_selection_artist.cpp:
```diff
@@ -0,0 +1,487 @@
+  marker_array_.markers = makeVisual(sensor_frame);
+}
+
+bool PathSelectionArtist::clearPathPointsCb(std_srvs::TriggerRequest& req, std_srvs::TriggerResponse& res)
+{
+  (void)req;  // To suppress warnings, tell the compiler we will not use this parameter
```
Prefer to delete name of the variable from the function signature instead
```c++
bool PathSelectionArtist::clearPathPointsCb(std_srvs::TriggerRequest&, std_srvs::TriggerResponse& res)
```

marip8 commented on opp_path_selection/src/path_selection_artist.cpp:
```diff
@@ -0,0 +1,487 @@
+      points_idx.push_back(idx);
+    }
+  
+  if(points_idx.size()>2) return true;
+  ROS_ERROR("No path points found");
+  return false;
```
Probably can simplify with;
```c++
ROS_ERROR_STREAM_COND(points_idx.size() > 2, "No path points found")
return points_idx.size() > 2;
```

marip8 commented on opp_path_selection/src/path_selection_artist.cpp:
```diff
@@ -0,0 +1,487 @@
+  pcl::PointCloud<pcl::PointXYZ>::Ptr cloud(new pcl::PointCloud<pcl::PointXYZ>());
+  pcl::fromROSMsg(req.input_cloud, *cloud);
+
+  // get parameters from file. Note: This file is specified on construction, re-read allows dynamic changes 
+  PathSelectorParameters params;
+  bool success = opp_msgs_serialization::deserialize(path_selection_config_file, params);
+  if (!success)
+  {
+    ROS_ERROR_STREAM("Could not load path selection config from: " << path_selection_config_file);
+    return false;
+  }
```
This isn't really a great practice. If you want this kind of dynamic behavior, these parameters should be part of the service request

marip8 commented on opp_path_selection/src/path_selection_artist.cpp:
```diff
@@ -0,0 +1,487 @@
+    points.push_back(std::move(e));
+  }
+
+  // get parameters from file. Note: This file is specified on construction, re-read allows dynamic changes 
+  PathSelectorParameters params;
+  bool success = opp_msgs_serialization::deserialize(path_selection_config_file, params);
+  if (!success)
+  {
+    ROS_ERROR_STREAM("Could not load path selection config from: " << path_selection_config_file);
+    return false;
+  }
```
Same comment about dynamic parameters being part of the service request instead

marip8 commented on opp_path_selection/src/path_selection_node.cpp:
```diff
@@ -0,0 +1,68 @@
+    return 1;
+  }
+
+  // Wait for tf to initialize
+  tf::TransformListener listener;
+  if (!listener.waitForTransform(
+          world_frame, sensor_data_frame, ros::Time(0), ros::Duration(static_cast<double>(TIMEOUT))))
+  {
+    ROS_ERROR("Transform lookup between '%s' and '%s' timed out", world_frame.c_str(), sensor_data_frame.c_str());
+    return -1;
+  }
```
Same comment about handling this failure appropriately when the transform is looked up rather than trying to preempt the failure

marip8 commented on opp_path_selection/src/path_selector.cpp:
```diff
@@ -0,0 +1,288 @@
+
+namespace opp_path_selection
+{
+int PathSelector::pointToLineSegmentDistance(const Eigen::Vector3d& segStart,
```
Why not return boolean? Seems like that would be more normal

marip8 commented on opp_path_selection/src/path_selector.cpp:
```diff
@@ -0,0 +1,288 @@
+  std::vector<int> point_seg_indices;
+  for(int i=0; i<points.size(); i++)
+    {
+      double mind = 2000.0;
```
```suggestion
      double min_distance = std::numeric_limits<double>::max();
```

marip8 replied:
2000 seems like an arbitrary value and the name is a bit confusing

marip8 commented on opp_path_selection/src/path_selector.cpp:
```diff
@@ -0,0 +1,288 @@
+      for(int j=0; j<input_mesh.vertices.size(); j++)
+	{
+	  Eigen::Vector3d pt2(input_mesh.vertices[j].x, input_mesh.vertices[j].y, input_mesh.vertices[j].z);
+	  Eigen::Vector3d D = pt2 - points[i];
```
A more descriptive name (or comment) would be helpful

marip8 commented on opp_path_selection/src/path_selector.cpp:
```diff
@@ -0,0 +1,288 @@
+				     index_map,		          // takes a vertex descriptor and finds the index of the vertex
+				     std::less<double>(),
+				     boost::closed_plus<double>(), 
+				     1000.0,
+				     //			      (std::numeric_limits<double>::max)(),
```
Why not use the numeric limits max?

marip8 commented on opp_path_selection/src/path_selector.cpp:
```diff
@@ -0,0 +1,288 @@
+
+  pcl::KdTreeFLANN<pcl::PointXYZ> kdtree;
+  kdtree.setInputCloud(input_cloud->makeShared());
+  int K = 8;
```
Seems arbitrary; should this be an argument to the function

marip8 commented on opp_path_selection/src/path_selector.cpp:
```diff
@@ -0,0 +1,288 @@
+  std::vector<int> point_seg_indices;
+  for(int i=0; i<points.size(); i++)
+    {
+      double mind = 2000.0;
```
Same comment as above

marip8 commented on opp_path_selection/src/path_selector.cpp:
```diff
@@ -0,0 +1,288 @@
+      size_t num_edges = boost::num_edges(G);
+      p.resize(num_verts);
+      d.resize(num_verts);
+      boost::property_map<MeshGraph, boost::edge_weight_t>::type weightmap = boost::get(boost::edge_weight, G);
+      boost::property_map<MeshGraph, boost::vertex_index_t>::type index_map = get(boost::vertex_index, G);
+      vertex_des start_vertex = vertex(point_seg_indices[i-1],G);
+      boost::dijkstra_shortest_paths(G,                           // const Graph& g
+				     start_vertex,	          // vertex_descriptor s, source, all shortest path distances computed from here
+				     &p[0],                       // predecessorMap each vertex p[i] points the next closer one p[p[i]] the next after
+				     &d[0],                       // distanceMap result providing the distance to each vertex from s
+				     weightmap,                   // weightMap accesses the weight of each edge
+				     index_map,		          // takes a vertex descriptor and finds the index of the vertex
+				     std::less<double>(),
+				     boost::closed_plus<double>(), 
+				     1000.0,
+				     //			      (std::numeric_limits<double>::max)(),
+				     0.0,
+				     boost::default_dijkstra_visitor());
+      
+      int V = point_seg_indices[i];
+      int start_index = point_seg_indices[i-1];
+      int q=0;
+      // follow predicesor chain from end endex to start index, pushing each one onto the 
+      std::vector<int> seg_indices;
+      while(V != start_index && q<p.size())
+	{
+	  seg_indices.push_back(V);
+	  V = p[V];
+	  q++;
+	}
+      // TODO check that q != p.size(), in this case, there is no path between the two points in the graph
+      seg_indices.push_back(V);
+
+      // reverse order of each segment
+      for(int j=seg_indices.size()-1; j>=0; j--)
+	{
+	  path_indices.push_back(seg_indices[j]);
+	}
+    }  // end of for each source path segment
+  return path_indices;
```
Seems like you could put this into a single helper function that can be used by both methods

marip8 commented on opp_area_selection/include/opp_area_selection/selection_artist.h:
```diff
@@ -56,7 +56,7 @@ class SelectionArtist
 
 protected:
-  void getSensorData(const sensor_msgs::PointCloud2::ConstPtr& msg);
+  //  void getSensorData(const sensor_msgs::PointCloud2::ConstPtr& msg); NOT IMPLEMENTED AND UNUSED
```
Remove if it's not implemented and unused

marip8 commented on opp_gui/src/widgets/polygon_area_selection_widget.cpp:
```diff
@@ -43,6 +43,7 @@ PolygonAreaSelectionWidget::~PolygonAreaSelectionWidget() { delete ui_; }
 
 void PolygonAreaSelectionWidget::init(const shape_msgs::Mesh& mesh)
 {
+  //  submesh_.reset(new shape_msgs::Mesh(*mesh_)); is also called inside clearROISelection(), is it needed here?
```
Remove commented code

marip8 commented on opp_gui/src/widgets/polyline_path_selection_widget.cpp:
```diff
@@ -0,0 +1,160 @@
+  return;
+}
+
+void PolylinePathSelectionWidget::writeMeshAsObj(const std::string& filename)
+{
+  ROS_ERROR("writing mesh to %s", filename.c_str());
+  FILE *fp = fopen(filename.c_str(),"w");
+  if(!fp)
+    {
+      ROS_ERROR("couldn't open file: %s", filename.c_str());
+    }
+  else
+    {
+      for(int i=0; i<mesh_->vertices.size(); i++)
+	{
+	  fprintf(fp, "v %lf %lf %lf\n", mesh_->vertices[i].x, mesh_->vertices[i].y, mesh_->vertices[i].z);
+	}
+      for(int i=0; i<mesh_->triangles.size(); i++)
+	{
+	  fprintf(fp, "f %ld %ld %ld\n",
+		  mesh_->triangles[i].vertex_indices[0]+1,   // for some reason, .obj files are 1 referenced, not zero referenced
+		  mesh_->triangles[i].vertex_indices[1]+1,
+		  mesh_->triangles[i].vertex_indices[2]+1);
+	}
+    }
+  fclose(fp);
+}
```
Consider using a [built-in PCL method](https://pointclouds.org/documentation/group__io.html#gabc0a54f1960ca03d5fa74fe3c63887a0) instead

drchrislewis replied:
This function is no longer used by the GUI. Perhaps we just remove it.

marip8 commented on opp_gui/src/widgets/polyline_path_selection_widget.cpp:
```diff
@@ -0,0 +1,160 @@
+    }
+  fclose(fp);
+}
+void PolylinePathSelectionWidget::writePolylineAsSource(const std::string& filename, const std::vector<int>& path_indices)
+{
+  FILE *fp = fopen(filename.c_str(),"w");
+  if(!fp)
+    {
+      ROS_ERROR("couldn't open file: %s", filename.c_str());
+    }
+  else
+    {
+      ROS_ERROR("source to %s", filename.c_str());
+      fprintf(fp,"1\n%ld ", path_indices.size());
+      for(int i=0; i<path_indices.size(); i++)
+	{
+	  fprintf(fp, "%ld ", path_indices[i]+1);
+	}
+      fprintf(fp,"\n");
+    }
+  fclose(fp);
+
+}
```
You can use YAML serialization directly for this instead of writing a custom function

drchrislewis replied:
Also no longer used. Remove?

marip8 commented on opp_gui/src/widgets/tool_path_parameters_editor_widget.cpp:
```diff
@@ -209,6 +312,65 @@ void ToolPathParametersEditorWidget::onGenerateToolPathsComplete(
+  if (state != actionlib::SimpleClientGoalState::SUCCEEDED)
+  {
+    std::string message = "Action '" + GENERATE_HEAT_TOOLPATHS_ACTION + "' failed to succeed";
+    QMessageBox::warning(this, "Tool Path Planning Error", QString(message.c_str()));
```
Creating a Qt message box in a function that is called by the ROS thread rather than the Qt thread (this function will be called by the action server when it completes) will crash your program due to "recursive painting". You can get around this by emitting a signal with your message and connecting it with a separate function in this class that creates the message dialog box. The trigger via signal makes sure that the message box is created in the Qt thread

drchrislewis replied:
Please take this one off line too. I don't understand you here either.

marip8 commented on opp_gui/src/widgets/tool_path_parameters_editor_widget.cpp:
```diff
@@ -209,6 +312,65 @@ void ToolPathParametersEditorWidget::onGenerateToolPathsComplete(
+  {
+    if (!res->success || !res->tool_path_validities[0])
+    {
+      QMessageBox::warning(this, "Heat Tool Path Planning Error", "Heat tool path generation failed");
```
Same comment as above

drchrislewis replied:
@marip8 Please call and we'll discuss the indicated items off line

Commit:
Partial satisfication of PR review comments

Commit:
forgot to save before last commit

Commit:
Created signal/slot for QMessageBox to avoid crashes from recursive repaint
Moved filter_impl.h to filter_impl.hpp and libs now only include filter_impl.hpp
Removed unused write to file functions
Removed unused service that was not even built in CMakelists.txt
moved as many headers from path_selector.h into path_selector.cpp as possible

drchrislewis commented:
@marip8 Please review may changes

marip8 commented on opp_gui/src/widgets/polyline_path_selection_widget.cpp:
```diff
@@ -0,0 +1,115 @@
+  if (!srv.response.success)
+  {
+    ROS_ERROR_STREAM("Tool Path Parameter Editor Widget: Area Selection error:" << srv.response.message);
+  }
```
You may want to use the `QMessageBox` here instead

marip8 commented on opp_gui/src/widgets/polyline_path_selection_widget.cpp:
```diff
@@ -0,0 +1,115 @@
+  bool success = selector_.collectPathMesh(*mesh_, path_indices, error_message);
+  if (!success)
+  {
+    ROS_ERROR_STREAM(
```
Use `QMessageBox` instead to communicate this error

marip8 commented on opp_gui/src/widgets/polyline_path_selection_widget.cpp:
```diff
@@ -0,0 +1,115 @@
+  bool success = selector_.collectPath(*mesh_, path_indices, error_message);
+  if (!success)
+  {
+    ROS_ERROR_STREAM(
```
Use `QMessageBox` to communicate error

marip8 commented on opp_gui/src/widgets/tool_path_editor_widget.cpp:
```diff
@@ -255,6 +255,16 @@ void ToolPathEditorWidget::publishToolPathDisplay(const opp_msgs::ToolPath& tool
+void ToolPathEditorWidget::onPolylinePath(const std::vector<int> pnt_indices)
+{
+  // First find shortest path on surface between each segments same as with pathGen
+  ROS_ERROR("new polyline path in ToolPathEditor has %ld pnts", pnt_indices.size());
```
Use `QMessageBox` to communicate error

marip8 commented on opp_gui/src/widgets/tool_path_parameters_editor_widget.cpp:
```diff
@@ -209,6 +316,74 @@ void ToolPathParametersEditorWidget::onGenerateToolPathsComplete(
+void ToolPathParametersEditorWidget::onQErrorBox(std::string error_string)
+{
+  QMessageBox::warning(this, "Tool Path Planning Error", QString(error_string.c_str()));
+}
```
Consider using a single function instead. I think `QMessageBox::warning` is the highest error level; there is also `QMessageBox::information` which you could use to communicate important information that is not an error

drchrislewis replied:
@marip8 One more time?

Commit:
replaced several ROS_ERRORs with QMessageBox()

Commit:
removed unnecessary onQErrorBox()

Commit:
added raster_angle to the config for the heat method.

Commit:
this code crashes when I add points. I've worked and worked and can't figure out why

Commit:
Code works, just needed to compile rviz from source and remove all my machinations trying to fix it

Commit:
just minor changes in ROS_INFOs

Commit:
removed commented out line

drchrislewis committed:
Merge pull request #2 from drchrislewis/stash_failure

Stash failure

Commit:
added checkboxes to both polyline and polygon editing to enable/disable adding points of each type independently

marip8 commented:
@drchrislewis please rebase this onto the current master branch when your changes are finished so the new CI will run
